### PR TITLE
Update egg to 0.8

### PR DESCRIPTION
**Pull request by mwillsey**
This updates egg to 0.8. Along the way, I made some testing things a little less brittle, so it accepts alpha-renamed rules as equivalent (see bool).

One big note: since 0.7, `apply_one` **performs a union**. So that broke our `NotDefined` applier. I fixed it by copy-pasting the code for applying (but not unioning) a pattern in egg. So we use that instead of `apply_one` in our implementation of `apply_one`.

### CODE REVIEW

**Review by chandrakananandi — COMMENTED** [review]
Ok so looks like the regression tests (bv4) failed due to reduced proving power. We can't derive this rule that we expect. 
`(<< (-- ?b ?b) ?a) => (-- ?a ?a)`

Maybe we should discuss why this is happening. A different but equally "powerful" ruleset is fine, but this seems odd.

File: src/equality.rs [hunk 1]
```diff
@@ -22,6 +22,43 @@ struct SerializedEq {
...
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if let Some((l, r)) = s.split_once("=>") {
+            Ok(Self {
+                lhs: l.into(),
+                rhs: r.into(),
+                bidirectional: false,
+            })
+        } else if let Some((r, l)) = s.split_once("<=") {
```
[the remark this answers is not in the document]
Comment (mwillsey) [reply src/equality.rs:35]: hmm, probably not. doesn't hurt, i guess.

**Comment by mwillsey** [comment]
Something seems off about the bv rules (both the new and the old ones). Where are the constants? Note that the missed rule ` (<< (-- ?b ?b) ?a) => (-- ?a ?a)` is just ` (<< 0 ?a) => 0`.

**Comment by chandrakananandi** [comment]
> Something seems off about the bv rules (both the new and the old ones). Where are the constants? Note that the missed rule ` (<< (-- ?b ?b) ?a) => (-- ?a ?a)` is just ` (<< 0 ?a) => 0`.

Well looking at the code, seems like we don't enumerate them! `init_synth` seems to have all the lines for adding the constants commented out. This was for the cvc4 eval as far as I remember.

**Comment by bksaiki** [comment]
Can't you find that rule by constant folding? Or am I missing something...

**Comment by chandrakananandi** [comment]
> Can't you find that rule by constant folding? Or am I missing something...

oh hmmm that's a good point..

**Comment by bksaiki** [comment]
@cnandi I investigated the `bv4` issues. I'd be okay with updating the unit test and merging. I ran a bi-directional derive test, and I got this result:
```
Using out.json to derive tests/bv4.json
46 rules are derivable, 1 are not.

Using tests/bv4.json to derive out.json
46 rules are derivable, 2 are not.
```
The rules that cannot be derived in either direction are due to variable naming differences:
 - `(<< (-- ?b ?b) ?a) => (-- ?a ?a)` (from current branch)
 - `(<< (-- ?b ?b) ?a) => (-- ?b ?b)` (baseline)

Interestingly, the current branch finds a new rule `(& ?a (<< ?b ?a)) => (-- ?b ?b)` that is underivable by our baseline rules, assuming it is valid.

**Comment by bksaiki** [comment]
Never mind, it's also due to variable naming: `(& ?a (<< ?b ?a)) => (-- ?a ?a)`.

**Comment by bksaiki** [comment]
Update: strange things are still occuring with rationals but the bv4 seems fine and its unit tests can be updated (if need be to be pass CI).

**Pull request merged**